### DISCUSSION TITLE
Reflecting Models

### DISCUSSION THREAD

**Discussion by jmarbutt** [prisma/prisma#11142] [Q&A] [answered]
Is there a way to reflect on the models from code? I want to read my model structure in javascript, is this possible?

**Comment by austincrim** [comment, marked as answer]
Hey @jmarbutt 👋 ,

We have an undocumented, unstable API for this, `Prisma.dmmf`. You can read it like this:
```ts
import { Prisma } from '@prisma/client'

console.log(Prisma.dmmf)
```

This will most likely have what you need, but keep in mind it is not documented or reflected in our semantic versioning strategy.

Let me know if you have any more questions!

**Reply by jmarbutt** [reply]
This is exactly what I am looking for!

**Reply by jmarbutt** [reply]
Hey @austincrim is there a way to put additional annotations on the schema model? Since Prisma doesn't use classes, it would be nice to allow some custom annotations. I want to be able to add a label for the field, I know this may be a stretch to ask for but I think it would be awesome since we have no other way around it in the Prisma schema.

**Reply by austincrim** [reply]
Hey @jmarbutt,

The closest thing I can think of would be a [triple-slash comment](https://www.prisma.io/docs/concepts/components/prisma-schema#comments) in your schema. Whatever you put in there will show up in the `dmmf`, however it's unstructured and might not work for you depending on your use case.

**Reply by a-type** [reply]
Just jumping in here - this would be really helpful for me, as I'm trying to convert some raw Postgres rows I get from Supabase subscriptions into Prisma model objects so they can be used by the rest of our system.

However, I'm not seeing any mapping of column names from Prisma's model name to the original database column. This is mostly relevant in our case for mapping db snake_case to camelCase, but we've also renamed a few fields outright (so I can't just run it through a case conversion).

Am I missing those field name mappings? Is there a more 'blessed' way to normalize raw database rows into Prisma models?

**Comment by samrathchadha** [comment]
>additional annotations on the schema model

What I usually use is 
```typescript
import { Model } from "@prisma/client"

type props {
  modelExtension: Model & { extraValue string, extraExtraValue: string[] };
}
```
I just define it globally/in a type.ts file and import it when needed

If you want the default prisma client to use that in your code(might break internal prisma features). You can overwrite it in a `global.d.ts` file by extending the prisma namespace and then redeclaring an interface extended from the Model(which you must import).

```typescript
// global.d.ts
import { Model } from "@prisma/client";

declare module "prisma" {
  interface Model extends Model {
    extraValue string;
    extraExtraValue: string[];
  }
}
```

If you use optional types(in the `global.d.ts` implementation) it might not break internal features but it may degrade type-safety

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
